README.md
# BARF : Binary Analysis and Reverse engineering Framework

[![Build Status](https://travis-ci.org/programa-stic/barf-project.svg?branch=master)](https://travis-ci.org/programa-stic/barf-project)

The analysis of binary code is a crucial activity in many areas of the computer sciences and software engineering disciplines ranging from software security and program analysis to reverse engineering. Manual binary analysis is a difficult and time-consuming task and there are software tools that seek to automate or assist human analysts. However, most of these tools have several technical and commercial restrictions that limit access and use by a large portion of the academic and practitioner communities. *BARF* is an open source binary analysis framework that aims to support a wide range of binary code analysis tasks that are common in the information security discipline. It is a scriptable platform that supports instruction lifting from multiple architectures, binary translation to an intermediate representation, an extensible framework for code analysis plugins and interoperation with external tools such as debuggers, SMT solvers and instrumentation tools. The framework is designed primarily for human-assisted analysis but it can be fully automated.

The *BARF project* includes *BARF* and related tools and packages. So far the
project is composed of the following items:

* **BARF** : A multiplatform open source Binary Analysis and Reverse engineering Framework
* **PyAsmJIT** : A JIT for the Intel x86_64 and ARM architecture.
* Tools built upon *BARF*:
    * **BARFgadgets** : Lets you *search*, *classifiy* and *verify* ROP gadgets inside a binary program.
    * **BARFcfg** : Lets you recover the control-flow graph of the functions of a binary program.
    * **BARFcg** : Lets you recover the call graph of the functions of a binary program.

For more information, see:

* *BARF: A multiplatform open source Binary Analysis and Reverse engineering Framework* (Whitepaper) [[en](./doc/papers/barf.pdf)]
* *BARFing Gadgets* (ekoparty2014 presentation) [[es](./doc/presentations/barfing-gadgets.ekoparty2014.es.pdf)]

Current status:

| **Latest Release** | v0.6.0                                                                 |
|-------------------:|:-----------------------------------------------------------------------|
|            **URL** | https://github.com/programa-stic/barf-project/releases/tag/v0.6.0      |
|     **Change Log** | https://github.com/programa-stic/barf-project/blob/v0.6.0/CHANGELOG.md |

> All packages were tested on Ubuntu 16.04 (x86_64).

## BARF

*BARF* is a Python package for binary analysis and reverse engineering. It can:

* Load binary programs in different formats (``ELF``, ``PE``, etc),
* It supports the Intel x86 architecture for 32 and 64 bits,
* It supports the ARM architecture for 32 bits,
* It operates on an intermediate language ([REIL]) thus all analysis algorithm are architecture-agnostic,
* It has integration with [Z3] and [CVC4] SMT solvers which means that you can express fragments of code as formulae and check restrictions on them.

It is currently *under development*.

### Installation

*BARF* depends on the following SMT solvers:

* [Z3] : A high-performance theorem prover being developed at Microsoft Research.
* [CVC4] : An efficient open-source automatic theorem prover for satisfiability modulo theories (SMT) problems.

The following command installs *BARF* on your system:

```bash
$ sudo python setup.py install
```

You can also install it locally:

```bash
$ sudo python setup.py install --user
```

#### Notes

* Only one SMT solver is needed in order to work. You may choose between Z3 and CVC4 or install both.
* To run some tests you need to install [PyAsmJIT] first: ``sudo pip install pyasmjit``
* You may need to install [Graphviz]: ``sudo apt-get install graphviz``

### Quickstart

This is a very simple example which shows how to open a binary file and print each instruction with its translation to the intermediate language (*REIL*).

```python
from barf import BARF

# Open binary file.
barf = BARF("examples/misc/samples/bin/branch4.x86")

# Print assembly instruction.
for addr, asm_instr, reil_instrs in barf.translate():
    print("{:#x} {}".format(addr, asm_instr))

    # Print REIL translation.
    for reil_instr in reil_instrs:
        print("\t{}".format(reil_instr))
```

We can also recover the CFG and save it to a ``.dot`` file.

```python
# Recover CFG.
cfg = barf.recover_cfg()

# Save CFG to a .dot file.
cfg.save("branch4.x86_cfg")
```

We can check restrictions on code using a SMT solver. For instance, suppose you have the following code:

```
 80483ed:       55                      push   ebp
 80483ee:       89 e5                   mov    ebp,esp
 80483f0:       83 ec 10                sub    esp,0x10
 80483f3:       8b 45 f8                mov    eax,DWORD PTR [ebp-0x8]
 80483f6:       8b 55 f4                mov    edx,DWORD PTR [ebp-0xc]
 80483f9:       01 d0                   add    eax,edx
 80483fb:       83 c0 05                add    eax,0x5
 80483fe:       89 45 fc                mov    DWORD PTR [ebp-0x4],eax
 8048401:       8b 45 fc                mov    eax,DWORD PTR [ebp-0x4]
 8048404:       c9                      leave
 8048405:       c3                      ret
```

And you want to know what values you have to assign to memory locations ``ebp-0x4``, ``ebp-0x8`` and ``ebp-0xc`` in order to obtain a specific value in ``eax`` register after executing the code.

First, we add the instructions to the analyzer component.

```python
from barf import BARF

# Open ELF file
barf = BARF("examples/misc/samples/bin/constraint1.x86")

# Add instructions to analyze.
for addr, asm_instr, reil_instrs in barf.translate(0x80483ed, 0x8048401):
    for reil_instr in reil_instrs:
        barf.code_analyzer.add_instruction(reil_instr)
```

Then, we generate expressions for each variable of interest and add the desired restrictions on them.

```python
ebp = barf.code_analyzer.get_register_expr("ebp", mode="post")

# Preconditions: set range for variable a and b
a = barf.code_analyzer.get_memory_expr(ebp-0x8, 4, mode="pre")
b = barf.code_analyzer.get_memory_expr(ebp-0xc, 4, mode="pre")

for constr in [a >= 2, a <= 100, b >= 2, b <= 100]:
    barf.code_analyzer.add_constraint(constr)

# Postconditions: set desired value for the result
c = barf.code_analyzer.get_memory_expr(ebp-0x4, 4, mode="post")

for constr in [c >= 26, c <= 28]:
    barf.code_analyzer.add_constraint(constr)
```

Finally, we check is the restrictions we establish can be resolved.

```python
if barf.code_analyzer.check() == 'sat':
    print("[+] Satisfiable! Possible assignments:")

    # Get concrete value for expressions
    a_val = barf.code_analyzer.get_expr_value(a)
    b_val = barf.code_analyzer.get_expr_value(b)
    c_val = barf.code_analyzer.get_expr_value(c)

    # Print values
    print("- a: {0:#010x} ({0})".format(a_val))
    print("- b: {0:#010x} ({0})".format(b_val))
    print("- c: {0:#010x} ({0})".format(c_val))

    assert a_val + b_val + 5 == c_val
else:
    print("[-] Unsatisfiable!")
```

You can see these and more examples in the [examples](./examples) directory.

### Overview

The framework is divided in three main components: **core**, **arch** and **analysis**.

#### Core

This component contains essential modules:

* ``REIL``: Provides definitions for the REIL language. It, also, implements an *emulator* and a *parser*.
* ``SMT``: Provides means to interface with [Z3] and [CVC4] SMT solver. Also, it provides functionality to translate REIL instructions to SMT expressions.
* ``BI``: The *Binary Interface* module is responsible for loading binary files for processing (it uses [PEFile] and [PyELFTools].)

#### Arch

Each supported architecture is provided as a subcomponent which contains the following modules.

* ``Architecture``: Describes the architecture, i.e., registers, memory address size.
* ``Translator``: Provides translators to REIL for each supported instruction.
* ``Disassembler``: Provides disassembling functionalities (it uses [Capstone].)
* ``Parser``: Transforms instruction from string to object form.

#### Analysis

So far this component consists of modules: *Control-Flow Graph*, *Call Graph* and *Code Analyzer*. The first two, provides functionality for CFG and CG recovery, respectively. The latter, its a high-level interface to the SMT solver related functionality.

## Tools

### BARFgadgets

``BARFgadgets`` is a Python script built upon BARF that lets you *search*, *classifiy* and *verify* ROP gadgets inside a binary program. The *search* stage finds all ``ret``-, ``jmp``- and ``call``-ended gadgets inside the binary. The *classification* stage classifies previously found gadgets according to the following types:

* No-Operation,
* Move Register,
* Load Constant,
* Arithmetic/Logical Operation,
* Load Memory,
* Store Memory,
* Arithmetic/Logical Load,
* Arithmetic/Logical Store and
* Undefined.

This is done through instruction emulation. Finally, the *verification* stage consists of using a SMT solver to verify the semantic assigned to each gadget in the second stage.

```
usage: BARFgadgets [-h] [--version] [--bdepth BDEPTH] [--idepth IDEPTH] [-u]
                   [-c] [-v] [-o OUTPUT] [-t] [--sort {addr,depth}] [--color]
                   [--show-binary] [--show-classification] [--show-invalid]
                   [--summary SUMMARY] [-r {8,16,32,64}]
                   filename

Tool for finding, classifying and verifying ROP gadgets.

positional arguments:
  filename              Binary file name.

optional arguments:
  -h, --help            show this help message and exit
  --version             Display version.
  --bdepth BDEPTH       Gadget depth in number of bytes.
  --idepth IDEPTH       Gadget depth in number of instructions.
  -u, --unique          Remove duplicate gadgets (in all steps).
  -c, --classify        Run gadgets classification.
  -v, --verify          Run gadgets verification (includes classification).
  -o OUTPUT, --output OUTPUT
                        Save output to file.
  -t, --time            Print time of each processing step.
  --sort {addr,depth}   Sort gadgets by address or depth (number of
                        instructions) in ascending order.
  --color               Format gadgets with ANSI color sequences, for output
                        in a 256-color terminal or console.
  --show-binary         Show binary code for each gadget.
  --show-classification
                        Show classification for each gadget.
  --show-invalid        Show invalid gadget, i.e., gadgets that were
                        classified but did not pass the verification process.
  --summary SUMMARY     Save summary to file.
  -r {8,16,32,64}       Filter verified gadgets by operands register size.
```

For more information, see [README](./barf/tools/gadgets/README.md).

### BARFcfg

``BARFcfg`` is a Python script built upon BARF that lets you recover the
control-flow graph of a binary program.

```
usage: BARFcfg [-h] [-s SYMBOL_FILE] [-f {txt,pdf,png,dot}] [-t]
               [-d OUTPUT_DIR] [-b] [--show-reil]
               [--immediate-format {hex,dec}] [-a | -r RECOVER]
               filename

Tool for recovering CFG of a binary.

positional arguments:
  filename              Binary file name.

optional arguments:
  -h, --help            show this help message and exit
  -s SYMBOL_FILE, --symbol-file SYMBOL_FILE
                        Load symbols from file.
  -f {txt,pdf,png,dot}, --format {txt,pdf,png,dot}
                        Output format.
  -t, --time            Print process time.
  -d OUTPUT_DIR, --output-dir OUTPUT_DIR
                        Output directory.
  -b, --brief           Brief output.
  --show-reil           Show REIL translation.
  --immediate-format {hex,dec}
                        Output format.
  -a, --recover-all     Recover all functions.
  -r RECOVER, --recover RECOVER
                        Recover specified functions by address (comma
                        separated).
```

### BARFcg

``BARFcg`` is a Python script built upon BARF that lets you recover the
call graph of a binary program.

```
usage: BARFcg [-h] [-s SYMBOL_FILE] [-f {pdf,png,dot}] [-t] [-a | -r RECOVER]
              filename

Tool for recovering CG of a binary.

positional arguments:
  filename              Binary file name.

optional arguments:
  -h, --help            show this help message and exit
  -s SYMBOL_FILE, --symbol-file SYMBOL_FILE
                        Load symbols from file.
  -f {pdf,png,dot}, --format {pdf,png,dot}
                        Output format.
  -t, --time            Print process time.
  -a, --recover-all     Recover all functions.
  -r RECOVER, --recover RECOVER
                        Recover specified functions by address (comma
                        separated).
```

## PyAsmJIT

*PyAsmJIT* is a Python package for x86_64/ARM assembly code generation and execution.


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.

## [Unreleased]
### Added

### Changed

### Deprecated

### Removed

### Fixed

### Security

## [0.6.0] - 2019-11-24
### Added

- Add Python 3 compatibility.
- Add classes for managing call conventions.
- Add hook support to the `Emulator` class.
- Add `cconv` (calling convention) module.
- Add `ReilCpu` tests.
- Add `replay` tool to replay x86 execution traces.
- Add `x86.trace` module.
- Add `x86.helper` module.
- Add kao's toy project solution to the `examples` folder.
- Add flare-on 2015 challenge #2 solution to the `examples` folder.
- Add basic symbolic execution support.
- Add extra methods to `ReilSequence` and `ReilContainer` classes.
- Improve hook support of the `Emulator` class.
- Add support for `SHLD` instruction.

### Changed

- Refactor ARM package.
- Refactor ARM tests.
- Refactor x86 package.
- Refactor x86 tests.
- Improve `emulate-binary` example script.
- Refactor `ReilBuilder` class.
- Refactor instruction translators (both Intel and ARM). Group translators by category and create a module for each one.
- Improve `Emulator` class.
- Refactor `ReilCpu` tests.
- Rename `bi` module to `binary`.
- Refactor `Disassembler` class.
- Refactor `ReilEmulatorTainter` class.
- Refactor `ReilCpu` class.
- Refactor `reil.helpers` module.
- Refactor `ReilContainer` class.
- Refactor `ReilBuilder` class.
- Refactor `reil.tainter` module.
- Refactor `ReilCpu` class.
- Refactor `reil.parser` test module.
- Refactor `reil.emulator` test module.
- Rename `basicblock` package (and tests) to `graphs`.
- Rename `gadget` package (and tests) to `gadgets`.
- Rename `reilparser` module to `reil.parser`.
- Refactor `reilemulator` module. Split module into submodules: `emulator.cpu`, `emulator.emulator`, `emulator.memory`, and `emulator.tainter`.
- Refactor `arch.emulator` module.

### Deprecated

### Removed

- Remove `DualInstruction` class.

### Fixed

- Fix `RSB` ARM instruction.
- Fix control-flow graph rendering.
- Fix `SHRD` translation.
- Fix size of effective address calculation for Intel architecture.
- Fix `SHL` and `SHR` instruction translator.

### Security

## [0.5.0] - 2017-12-18
### Added
- Add architecture emulator class.
- Add support for Travis CI.
- Add tests for the `smt` package.
- Add svg ouput format for `BARFcfg` and `BARFcg` tools.
- Add `Dockerfile`.
- Add support for x86 instructions: `LAHF`, `XADD`.
- Add support for x86 sse instructions: `LDDQU`, `MOVAPS`, `MOVSD`.

### Changed
- Restructure `examples` directory and remove redundant examples.
- Restructure `tools` directory and move it into `barf` package.
- Overall code quality improvement in most modules.
- Revamp `smt` package.
- Refactor `codeanalyzer` module.
- Improve code quality of `basicblock` module.
- Restructure binary sample directory.
- Load all sections of a binary into memory by default.
- Update `ARM` architecural information.
- Refactor `emulate` method to support `x86_64`, `ARM` and `Thumb` code.

### Fixed
- Add `BAL` and `BGT` to the list of ARM branch instructions.
- Fix Capstone installation issues.
- Various fixes in the `smt` package.

### Removed
- Remove `translation_mode` parameter from x86/ARM translators.
- Remove deprecated `barf-install-solver.sh` script.
- Remove `smtlibv2.py` module dependency from `PySymEmu`.

## [0.4.0] - 2017-04-10
### Added
- Update `README` to include new command line options for all BARF tools.
- Add initial support for x86 `AVX` instruction set.
- Add support for x86 instructions: `BSF`, `BSWAP`, `BTS`, `CDQ`, `MOVSXD` and `SHRD`.
- Add support for x86 SSE instructions: `MOVD`, `MOVDQA`, `MOVDQU`, `MOVHPD`, `MOVLPD`, `MOVQ`, `PCMPEQB`, `PMINUB`, `PMOVMSKB`, `POR`, `PSHUFD`, `PSLLDQ`, `PSRLDQ`, `PSUBB`, `PUNPCKLBW`, `PUNPCKLWD`, `PUNPCKLQDQ` and `PXOR`.
- Add initial support for x86 `SSE` instruction set.
- Add support for `pdf`, `png` and `dot` output formats in `BARFcfg` and `BARFcg` tools.
- Add option to display immediate operand values in `hex` and `dec` to the CFG rendering.
- Add REIL instruction index to the CFG rendering.
- Add support for missing x86 flag translations: `AF` and `PF`.
- Add new method to the `ArchitectureInformation` class to retrieve information about syscall instructions.
- Add new x86 example for finding and executing functions.
- Add new methods to the `CallGraph` class.
- Add initial support for `GS` and `FS` segments access (x86).

### Changed
- Improve coding style and code quality of the `reil` module.
- Improve coding style and code quality of the `reilemulator` module.
- Remove SMT requirement (an exception is raised when trying to use related functionality).
- Improve coding style and code quality of the `x86` package.
- Improve `emulate_full` method to support any piece of code.
- Refactor `_open_elf` logic in the `BinaryFile` class by @Seraphime.

### Fixed
- Fix missing check on `recover_cfg` function parameters.
- Update missing branch instructions in the `X86ArchitectureInformation` class.
- Fix x86 instruction translation: `SAR`, `SHR`, `SHL`, `ROR`, `MOVZX` and `MOV`.
- Fix x86 gadget finding function.
- Fix various typos.
- Fix x86 `REP` prefix parsing.

## [0.3] - 2016-12-13
### Added
- Add new BARF tool, `BARFcg`, for CG recovery.
- Add support for CG recovery (x86 and ARM).
- Add new BARF tool, `BARFcfg`, for CFG recovery.
- Add support for ARM CFG recovery.
- Add support for more ARM instructions.
- Add support for data tainting in `ReilEmulator`.
- Add support for pre/post instruction execution callback function in `ReilEmulator`.
- Add support for REIL extension instruction (`SEXT`, `SDIV`, `SMOD`).
- Add support for more x86 instructions.

### Changed
- Improve CFG recovery functionality.
- Refactor `basickblock` module.
- Replace PyBFD with PyELFTools.
- Improve SMT performace.
- Refactor `smtlibv2` module.
- Overall directory restructure.
- Improve ARM disassembly integration to Capstone Engine.

reil.py
# Copyright (c) 2014, Fundacion Dr. Manuel Sadosky
# All rights reserved.

# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are met:

# 1. Redistributions of source code must retain the above copyright notice, this
# list of conditions and the following disclaimer.

# 2. Redistributions in binary form must reproduce the above copyright notice,
# this list of conditions and the following disclaimer in the documentation
# and/or other materials provided with the distribution.

# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

"""
This module contains all the classes that handle the intermediate
representation language. It is basically the REIL language with minor
changes. Below there is an overview of the REIL language and its
instruction format. For full details see "REIL: A platform-independent
intermediate representation of disassembled code for static code
analysis."

All algorithms within the framework are designed to operate on the
intermediate representation. This provides great flexibility when it
comes to implement a cross-platform framework.

Instruction Format
------------------

    mnemonic oprnd1, oprnd2, oprnd3

Instructions
------------

    Arithmetic    : ADD, SUB, MUL, DIV, MOD, BSH
    Bitwise       : AND, OR, XOR
    Data Transfer : LDM, STM, STR
    Conditional   : BISZ, JCC
    Other         : UNDEF, UNKN, NOP

"""

# Display operands size in intruction
show_size = True

class ReilMnemonic(object):

    """Enumeration of IR mnemonics.
    """

    # Arithmetic Instructions
    ADD   = 1
    SUB   = 2
    MUL   = 3
    DIV   = 4
    MOD   = 5
    BSH   = 6

    # Bitwise Instructions
    AND   = 7
    OR    = 8
    XOR   = 9

    # Data Transfer Instructions
    LDM   = 10
    STM   = 11
    STR   = 12

    # Conditional Instructions
    BISZ  = 13
    JCC   = 14

    # Other Instructions
    UNKN  = 15
    UNDEF = 16
    NOP   = 17

    # Added Instructions
    RET   = 18

    # Extensions
    SEXT  = 19

    @staticmethod
    def to_string(mnemonic):
        """Return the string representation of the given mnemonic.
        """
        strings = {
            # Arithmetic Instructions
            ReilMnemonic.ADD : "add",
            ReilMnemonic.SUB : "sub",
            ReilMnemonic.MUL : "mul",
            ReilMnemonic.DIV : "div",
            ReilMnemonic.MOD : "mod",
            ReilMnemonic.BSH : "bsh",

            # Bitwise Instructions
            ReilMnemonic.AND : "and",
            ReilMnemonic.OR : "or",
            ReilMnemonic.XOR : "xor",

            # Data Transfer Instructions
            ReilMnemonic.LDM : "ldm",
            ReilMnemonic.STM : "stm",
            ReilMnemonic.STR : "str",

            # Conditional Instructions
            ReilMnemonic.BISZ : "bisz",
            ReilMnemonic.JCC : "jcc",

            # Other Instructions
            ReilMnemonic.UNKN  : "unkn" ,
            ReilMnemonic.UNDEF : "undef" ,
            ReilMnemonic.NOP : "nop" ,

            # Added Instructions
            ReilMnemonic.RET : "ret",

            # Extensions
            ReilMnemonic.SEXT : "sext",
        }

        return strings[mnemonic]

    @staticmethod
    def from_string(string):
        """Return the mnemonic represented by the given string.
        """
        mnemonics = {
            # Arithmetic Instructions
            "add" : ReilMnemonic.ADD,
            "sub" : ReilMnemonic.SUB,
            "mul" : ReilMnemonic.MUL,
            "div" : ReilMnemonic.DIV,
            "mod" : ReilMnemonic.MOD,
            "bsh" : ReilMnemonic.BSH,

            # Bitwise Instructions
            "and" : ReilMnemonic.AND,
            "or" : ReilMnemonic.OR,
            "xor" : ReilMnemonic.XOR,

            # Data Transfer Instructions
            "ldm" : ReilMnemonic.LDM,
            "stm" : ReilMnemonic.STM,
            "str" : ReilMnemonic.STR,

            # Conditional Instructions
            "bisz" : ReilMnemonic.BISZ,
            "jcc" : ReilMnemonic.JCC,

            # Other Instructions
            "unkn" : ReilMnemonic.UNKN,
            "undef" : ReilMnemonic.UNDEF,
            "nop" : ReilMnemonic.NOP,

            # Added Instructions
            "ret" : ReilMnemonic.RET,

            # Added Instructions
            "sext" : ReilMnemonic.SEXT,
        }

        return mnemonics[string]


REIL_MNEMONICS = (
    # Arithmetic Instructions
    ReilMnemonic.ADD,
    ReilMnemonic.SUB,
    ReilMnemonic.MUL,
    ReilMnemonic.DIV,
    ReilMnemonic.MOD,
    ReilMnemonic.BSH,

    # Bitwise Instructions
    ReilMnemonic.AND,
    ReilMnemonic.OR,
    ReilMnemonic.XOR,

    # Data Transfer Instructions
    ReilMnemonic.LDM,
    ReilMnemonic.STM,
    ReilMnemonic.STR,

    # Conditional Instructions
    ReilMnemonic.BISZ,
    ReilMnemonic.JCC,

    # Other Instructions
    ReilMnemonic.UNKN,
    ReilMnemonic.UNDEF,
    ReilMnemonic.NOP,

    # Added Instructions
    ReilMnemonic.RET,

    # Extensions
    ReilMnemonic.SEXT,
)

class ReilInstruction(object):

    """Representation of a REIL instruction.
    """

    __slots__ = [
        '_mnemonic',
        '_operands',
        '_comment',
        '_address',
    ]

    def __init__(self):

        # A REIL mnemonic
        self._mnemonic = None

        # A list of operand. Exactly 3.
        self._operands = [ReilEmptyOperand()] * 3

        # Optionally, a comment for the instruction.
        self._comment = None

        # A REIL address for the instruction.
        self._address = None

    @property
    def mnemonic(self):
        """Get instruction mnemonic.
        """
        return self._mnemonic

    @property
    def mnemonic_str(self):
        """Get instruction mnemonic as string.
        """
        return ReilMnemonic.to_string(self._mnemonic)

    @mnemonic.setter
    def mnemonic(self, value):
        """Set instruction mnemonic.
        """
        if value not in REIL_MNEMONICS:
            raise Exception("Invalid instruction mnemonic : %s" % str(value))

        self._mnemonic = value

    @property
    def operands(self):
        """Get instruction operands.
        """
        return self._operands

    @operands.setter
    def operands(self, value):
        """Set instruction operands.
        """
        if len(value) != 3:
            raise Exception("Invalid instruction operands : %s" % str(value))

        self._operands = value

    @property
    def address(self):
        """Get instruction address.
        """
        return self._address

    @address.setter
    def address(self, value):
        """Set instruction address.
        """
        self._address = value

    @property
    def comment(self):
        """Get instruction comment.
        """
        return self._comment

    @comment.setter
    def comment(self, value):
        """Set instruction comment.
        """
        self._comment = value

    def __str__(self):
        def print_oprnd(oprnd):
            oprnd_str = str(oprnd)
            size_str = str(oprnd.size) if oprnd.size else ""

            sizes = {
                256 : "DDQWORD",
                128 : "DQWORD",
                72  : "POINTER",
                64  : "QWORD",
                40  : "POINTER",
                32  : "DWORD",
                16  : "WORD",
                8   : "BYTE",
                1   : "BIT",
                ""  : "UNK",
            }

            if isinstance(oprnd, ReilEmptyOperand):
                return "%s" % (oprnd_str)
            else:
                return "%s %s" % (sizes[oprnd.size if oprnd.size else ""], oprnd_str)

        mnemonic_str = ReilMnemonic.to_string(self._mnemonic)

        if show_size:
            operands_str = ", ".join(map(print_oprnd, self._operands))
        else:
            operands_str = ", ".join(map(str, self._operands))

        return "%-5s [%s]" % (mnemonic_str, operands_str)

    def __hash__(self):
        return hash(str(self))


class ReilOperand(object):

    """Representation of an IR instruction's operand.
    """

    __slots__ = [
        '_size',
    ]

    def __init__(self, size):

        # Size of the operand, in bits.
        self._size = size

    @property
    def size(self):
        """Get operand size.
        """
        return self._size

    @size.setter
    def size(self, value):
        """Set operand size.
        """
        self._size = value

    def __eq__(self, other):
        return  type(other) is type(self) and \
                self._size == other._size

    def __ne__(self, other):
        return not self.__eq__(other)


class ReilImmediateOperand(ReilOperand):

    """Representation of a REIL instruction immediate operand.
    """

    __slots__ = [
        '_immediate',
    ]

    def __init__(self, immediate, size=None):
        super(ReilImmediateOperand, self).__init__(size)

        assert type(immediate) in [int, long], "Invalid immediate value type."

        self._immediate = immediate

    @property
    def immediate(self):
        """Get immediate.
        """
        if not self._size:
            raise Exception("Operand size missing.")

        return self._immediate & 2**self._size-1

    def __str__(self):
        if not self._size:
            raise Exception("Operand size missing.")

        string = hex(self._immediate & 2**self._size-1)

        return string[:-1] if string[-1] == 'L' else string

    def __eq__(self, other):
        return  type(other) is type(self) and \
                self._size == other._size and \
                self._immediate == other._immediate


class ReilRegisterOperand(ReilOperand):

    """Representation of a REIL instruction register operand.
    """

    __slots__ = [
        '_name',
    ]

    def __init__(self, name, size=None):
        super(ReilRegisterOperand, self).__init__(size)

        # Register name.
        self._name = name

    @property
    def name(self):
        """Get IR register operand name.
        """
        return self._name

    def __str__(self):
        return self._name

    def __eq__(self, other):
        return  type(other) is type(self) and \
                self._size == other._size and \
                self._name == other._name


class ReilEmptyOperand(ReilRegisterOperand):

    """Representation of an IR instruction's empty operand.
    """

    def __init__(self):
        super(ReilEmptyOperand, self).__init__("EMPTY", size=None)


class ReilInstructionBuilder(object):

    """REIL Instruction Builder. Generate REIL instructions, easily.
    """

    # Arithmetic Instructions
    # ======================================================================== #
    def gen_add(self, src1, src2, dst):
        """Return an ADD instruction.
        """
        return self.build(ReilMnemonic.ADD, src1, src2, dst)

    def gen_sub(self, src1, src2, dst):
        """Return a SUB instruction.
        """
        return self.build(ReilMnemonic.SUB, src1, src2, dst)

    def gen_mul(self, src1, src2, dst):
        """Return a MUL instruction.
        """
        return self.build(ReilMnemonic.MUL, src1, src2, dst)

    def gen_div(self, src1, src2, dst):
        """Return a DIV instruction.
        """
        return self.build(ReilMnemonic.DIV, src1, src2, dst)

    def gen_mod(self, src1, src2, dst):
        """Return a MOD instruction.
        """
        return self.build(ReilMnemonic.MOD, src1, src2, dst)

    def gen_bsh(self, src1, src2, dst):
        """Return a BSH instruction.
        """
        return self.build(ReilMnemonic.BSH, src1, src2, dst)

    # Bitwise Instructions
    # ======================================================================== #
    def gen_and(self, src1, src2, dst):
        """Return an AND instruction.
        """
        return self.build(ReilMnemonic.AND, src1, src2, dst)

    def gen_or(self, src1, src2, dst):
        """Return an OR instruction.
        """
        return self.build(ReilMnemonic.OR, src1, src2, dst)

    def gen_xor(self, src1, src2, dst):
        """Return a XOR instruction.
        """
        return self.build(ReilMnemonic.XOR, src1, src2, dst)

    # Data Transfer Instructions
    # ======================================================================== #
    def gen_ldm(self, src, dst):
        """Return a LDM instruction.
        """
        return self.build(ReilMnemonic.LDM, src, ReilEmptyOperand(), dst)

    def gen_stm(self, src, dst):
        """Return a STM instruction.
        """
        return self.build(ReilMnemonic.STM, src, ReilEmptyOperand(), dst)

    def gen_str(self, src, dst):
        """Return a STR instruction.
        """
        return self.build(ReilMnemonic.STR, src, ReilEmptyOperand(), dst)

    # Conditional Instructions
    # ======================================================================== #
    def gen_bisz(self, src, dst):
        """Return a BISZ instruction.
        """
        return self.build(ReilMnemonic.BISZ, src, ReilEmptyOperand(), dst)

    def gen_jcc(self, src, dst):
        """Return a JCC instruction.
        """
        return self.build(ReilMnemonic.JCC, src, ReilEmptyOperand(), dst)

    # Other Instructions
    # ======================================================================== #
    def gen_unkn(self):
        """Return an UNKN instruction.
        """
        empty_reg = ReilEmptyOperand()

        return self.build(ReilMnemonic.UNKN, empty_reg, empty_reg, empty_reg)

    def gen_undef(self):
        """Return an UNDEF instruction.
        """
        empty_reg = ReilEmptyOperand()

        return self.build(ReilMnemonic.UNDEF, empty_reg, empty_reg, empty_reg)

    def gen_nop(self):
        """Return a NOP instruction.
        """
        empty_reg = ReilEmptyOperand()

        return self.build(ReilMnemonic.NOP, empty_reg, empty_reg, empty_reg)

    # Ad hoc Instructions
    # ======================================================================== #
    def gen_ret(self):
        """Return a RET instruction.
        """
        empty_reg = ReilEmptyOperand()

        return self.build(ReilMnemonic.RET, empty_reg, empty_reg, empty_reg)

    # Extensions
    # ======================================================================== #
    def gen_sext(self, src, dst):
        """Return a SEXT instruction.
        """
        empty_reg = ReilEmptyOperand()

        return self.build(ReilMnemonic.SEXT, src, empty_reg, dst)

    # Auxiliary functions
    # ======================================================================== #
    def build(self, mnemonic, oprnd1, oprnd2, oprnd3):
        """Return the specified instruction.
        """
        ins = ReilInstruction()

        ins.mnemonic = mnemonic
        ins.operands = [oprnd1, oprnd2, oprnd3]

        return ins


class DualInstruction(object):

    """Represents an assembler instruction paired with its IR
    representation.

    """

    __slots__ = [
        '_address',
        '_asm_instr',
        '_ir_instrs',
    ]

    def __init__(self, address, asm_instr, ir_instrs):

        # Address of the assembler instruction.
        self._address = address

        # Assembler instruction.
        self._asm_instr = asm_instr

        # REIL translation of the assembler instruction. Note that one
        # assemlber instruction is mapped to more than one REIL
        # instruction.
        self._ir_instrs = ir_instrs

    @property
    def address(self):
        """Get instruction address.
        """
        return self._address

    @property
    def asm_instr(self):
        """Get assembly instruction.
        """
        return self._asm_instr

    @property
    def ir_instrs(self):
        """Get IR representation of the assembly instruction.
        """
        return self._ir_instrs

    def __eq__(self, other):
        return  self.address == other.address and \
                self.asm_instr == other.asm_instr

    def __ne__(self, other):
        return not self.__eq__(other)
